tempesta_fw/http_sticky.c

```diff
@@ -0,0 +1,740 @@
+		} else  {
+			/*
+			 * Per RFC 1035, 2181, max length of FQDN is 255.
+			 * What if it is UTF-8 encoded?
+			 */
+			tfw_str_to_cstr(hdr + SLEN(S_F_HOST),
+					buf, hdr->len - SLEN(S_F_HOST));
+			tfw_http_msg_add_data(&state, msg, buf,
+					      hdr->len - SLEN(S_F_HOST));
+		}
+	}
+	TFW_STR_FOR_EACH_CHUNK(chunk, &req->uri_path) {
+		tfw_http_msg_add_data(&state, msg, chunk->ptr, chunk->len);
+	}
+	tfw_http_msg_add_data(&state, msg, S_CRLF, SLEN(S_CRLF));
+
+	tfw_http_msg_add_data(&state, msg, S_F_SET_COOKIE,
+					   SLEN(S_F_SET_COOKIE));
+	tfw_http_msg_add_data(&state, msg, tfw_cfg_sticky.name.ptr,
+					   tfw_cfg_sticky.name.len);
+	tfw_http_msg_add_data(&state, msg, "=", 1);
+	len = tfw_http_prep_sticky(buf, client->hmac, sizeof(client->hmac));
+	tfw_http_msg_add_data(&state, msg, buf, len);
+	tfw_http_msg_add_data(&state, msg, S_CRLFCRLF, SLEN(S_CRLFCRLF));
+
+bh_unlock_sock(client->sock); /* TEMPORARY TO SEE IT WORKING */
+	tfw_connection_send_cli(conn, (TfwMsg *)resp);
+	tfw_http_msg_free((TfwHttpMsg *)resp);
+
+	return 0;
+}
+
+#define	S_502_FIXLEN							\
+	SLEN(S_502)							\
+	+ SLEN(S_CRLF)							\
+	+ SLEN(S_F_DATE) + SLEN(S_V_DATE)				\
+	+ SLEN(S_CRLF)							\
+	+ SLEN(S_F_CONTENT_LENGTH) + SLEN(S_V_CONTENT_LENGTH)		\
+	+ SLEN(S_CRLFCRLF)
+/*
+ * Send an HTTP 502 response to the client. It tells the client that
+ * Tempesta is unable to forward the request to the designated server.
+ */
+int
+tfw_http_send_502(TfwHttpMsg *hm)
+{
+	u_char buf[256];
+	TfwMsg *msg;
+	TfwHttpResp *resp;
+	tfw_mastate_t state = { 0 };
+	size_t len, data_len = S_502_FIXLEN;
+	TfwConnection *conn = hm->conn;
+
+	if ((resp = (TfwHttpResp *) tfw_http_msg_alloc(Conn_Srv)) == NULL) {
+		return -1;
+	}
+	msg = (TfwMsg *)resp;
+
+	if (tfw_http_msg_setup(&state, msg, data_len) != 0) {
+		tfw_http_msg_free((TfwHttpMsg *)resp);
+		return -1;
+	}
+
+	tfw_http_msg_add_data(&state, msg, S_502, SLEN(S_502));
+	tfw_http_msg_add_data(&state, msg, S_CRLF, SLEN(S_CRLF));
+
+	tfw_http_msg_add_data(&state, msg, S_F_DATE, SLEN(S_F_DATE));
+	len = tfw_http_prep_date(buf);
+	tfw_http_msg_add_data(&state, msg, buf, len);
+	tfw_http_msg_add_data(&state, msg, S_CRLF, SLEN(S_CRLF));
+
+	tfw_http_msg_add_data(&state, msg, S_F_CONTENT_LENGTH,
+					   SLEN(S_F_CONTENT_LENGTH));
+	tfw_http_msg_add_data(&state, msg, "0", 1);
+	tfw_http_msg_add_data(&state, msg, S_CRLFCRLF, SLEN(S_CRLFCRLF));
+
+	tfw_connection_send_cli(conn, (TfwMsg *)resp);
+	tfw_http_msg_free((TfwHttpMsg *)resp);
+
+	return 0;
+}
+
+static void
+tfw_http_strim(tfw_str_t *s)
+{
+	u_char *sptr, *eptr;
+
+	if (!s->len)
+		return;
+
+	eptr = s->ptr + s->len - 1;
+	while (eptr >= s->ptr && isspace(*eptr))
+		eptr--;
+	s->len = eptr - s->ptr + 1;
+
+	sptr = skip_spaces(s->ptr);
+	s->len -= sptr - s->ptr;
+	s->ptr = sptr;
+}
+
+static TfwStr *
+tfw_http_field(TfwHttpMsg *hm, const char *field, size_t len)
+{
+	int i;
+	TfwStr *hdr;
+
+	for (i = 0; i < hm->h_tbl->size; i++) {
```

Comment:
Yes, that would be beneficial. But I also use this function to search for `User-Agent: ` header field. Would `User-Agent: ` header field be a good candidate for being a special header as well? That would be great.